File: killrvideo-commons/src/main/java/com/killrvideo/messaging/conf/KillrVideoThreadFactory.java
package com.killrvideo.messaging.conf;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.validation.constraints.NotNull;

/**
 * Custom ThreadFactory.
 *
 * @author DataStax Developer Advocates team.
 */
public class KillrVideoThreadFactory implements ThreadFactory {

	/** Create dedicated logger to trace ERRORS. */
    private static final Logger LOGGER = LoggerFactory.getLogger("killrvideo-default-executor");
    
    /** Counter keeping track of thread number in executor. */
    private final AtomicInteger threadNumber = new AtomicInteger(10);
    
    /**
     * Default constructor required for reflection.
     */
    public KillrVideoThreadFactory() {
    }

    /** {@inheritDoc} */
    public Thread newThread(@NotNull Runnable r) {
        Thread thread = new Thread(r);
        thread.setName("killrvideo-default-executor-" + this.threadNumber.incrementAndGet());
        thread.setDaemon(true);
        thread.setUncaughtExceptionHandler(this.uncaughtExceptionHandler);
        return thread;
    }
    
    /**
     * Overriding error handling providing logging.
     */
    private final Thread.UncaughtExceptionHandler uncaughtExceptionHandler = (t, e) ->
            LOGGER.error("Uncaught asynchronous exception : " + e.getMessage(), e);
}

File: killrvideo-commons/src/main/kotlin/com/killrvideo/messaging/conf/KillrVideoThreadFactory.kt
package com.killrvideo.messaging.conf

import org.slf4j.LoggerFactory
import java.lang.Thread.UncaughtExceptionHandler
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger

/**
 * Custom ThreadFactory.
 *
 * @author DataStax Developer Advocates team.
 */
class KillrVideoThreadFactory
/**
 * Default constructor required for reflection.
 */
    : ThreadFactory {
    /** Counter keeping track of thread number in executor.  */
    private val threadNumber = AtomicInteger(10)

    /** {@inheritDoc}  */
    override fun newThread(r: Runnable): Thread {
        val thread = Thread(r)
        thread.name = "killrvideo-default-executor-" + threadNumber.incrementAndGet()
        thread.isDaemon = true
        thread.uncaughtExceptionHandler = uncaughtExceptionHandler
        return thread
    }

    /**
     * Overriding error handling providing logging.
     */
    private val uncaughtExceptionHandler =
        UncaughtExceptionHandler { _: Thread, e: Throwable ->
            LOGGER.error("Uncaught asynchronous exception : " + e.message, e)
        }

    companion object {
        /** Create dedicated logger to trace ERRORS.  */
        private val LOGGER = LoggerFactory.getLogger("killrvideo-default-executor")
    }
}